### ISSUE TITLE
siad 1.3.1 host high io

### ISSUE THREAD

**Issue by chargr** [NebulousLabs/Sia#2590]
Bug Report
==========
I've noticed very high IO generated from siad, which is _probably_ amplified that I currently have my data under a btrfs file system (raid1).  The filesystem is generating about 250 write ops/sec (to a single device).

**Expected Behavior**
Less IO when hosting is "idle".

**Environment**
Main net

* Sia version:
```
$ siac version
Sia Client v1.3.1
Sia Daemon v1.3.1
```

* OS: 
`Linux desktop 4.14.8-1-ARCH #1 SMP PREEMPT Wed Dec 20 21:27:44 UTC 2017 x86_64 GNU/Linux`

`siad -M gctwh -d /data/sia/siad`

```
[sia@desktop sia]$ find /data/sia/ -mmin -1
/data/sia/siad/host/contractmanager
/data/sia/siad/host/contractmanager/contractmanager.json
/data/sia/siad/host/contractmanager/contractmanager.wal
/data/sia/siad/host/contractmanager/contractmanager.json_temp
/data/sia/siad/host/contractmanager/contractmanager.wal_temp
/data/sia/siad/transactionpool/transactionpool.db
/data/sia/siad/wallet/wallet.db
[sia@desktop sia]$ find /data/sia/ -mmin -5
/data/sia/siad/gateway/nodes.json
/data/sia/siad/gateway/nodes.json_temp
/data/sia/siad/host/contractmanager
/data/sia/siad/host/contractmanager/contractmanager.json
/data/sia/siad/host/contractmanager/contractmanager.wal
/data/sia/siad/host/contractmanager/contractmanager.wal_temp
/data/sia/siad/host/contractmanager/contractmanager.json_temp
/data/sia/siad/transactionpool/transactionpool.db
/data/sia/siad/wallet/wallet.db
```

I see in strace is that contractmanager is being rewritten frequently, less than a second apart. (no changes to contracts)  This shouldnt really account for all the IO I see.

```
$ time sudo strace -f -p 28832 2>&1 | grep -E '(open|rename)at'
[pid 28889] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json" <unfinished ...>
[pid 28889] <... renameat resumed> )    = 0
[pid 28889] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal") = 0
[pid 28889] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28883] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28889] <... openat resumed> )      = 8
[pid 28883] <... openat resumed> )      = 15
[pid 28839] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json" <unfinished ...>
[pid 28839] <... renameat resumed> )    = 0
[pid 28839] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal") = 0
[pid 28839] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28839] <... openat resumed> )      = 8
[pid 28838] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28838] <... openat resumed> )      = 15
[pid 28838] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json" <unfinished ...>
[pid 28838] <... renameat resumed> )    = 0
[pid 28838] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal" <unfinished ...>
[pid 28838] <... renameat resumed> )    = 0
[pid 28838] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666) = 8
[pid 28839] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28839] <... openat resumed> )      = 15
[pid 28834] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json" <unfinished ...>
[pid 28834] <... renameat resumed> )    = 0
[pid 28834] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal" <unfinished ...>
[pid 28834] <... renameat resumed> )    = 0
[pid 28834] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28834] <... openat resumed> )      = 8
[pid 28838] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28838] <... openat resumed> )      = 15
[pid 28839] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json" <unfinished ...>
[pid 28839] <... renameat resumed> )    = 0
[pid 28839] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal" <unfinished ...>
[pid 28839] <... renameat resumed> )    = 0
[pid 28839] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28839] <... openat resumed> )      = 8
[pid 28889] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28889] <... openat resumed> )      = 15
[pid 28883] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json" <unfinished ...>
[pid 28883] <... renameat resumed> )    = 0
[pid 28883] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal" <unfinished ...>
[pid 28883] <... renameat resumed> )    = 0
[pid 28883] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28838] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28883] <... openat resumed> )      = 8
[pid 28838] <... openat resumed> )      = 15
[pid 28836] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json" <unfinished ...>
[pid 28836] <... renameat resumed> )    = 0
[pid 28836] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal" <unfinished ...>
[pid 28836] <... renameat resumed> )    = 0
[pid 28836] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28836] <... openat resumed> )      = 8
[pid 28839] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28839] <... openat resumed> )      = 15
[pid 28839] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json" <unfinished ...>
[pid 28839] <... renameat resumed> )    = 0
[pid 28839] renameat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal") = 0
[pid 28839] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.wal_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28836] openat(AT_FDCWD, "/data/sia/siad/host/contractmanager/contractmanager.json_temp", O_RDWR|O_CREAT|O_TRUNC|O_CLOEXEC, 0666 <unfinished ...>
[pid 28839] <... openat resumed> )      = 8
[pid 28836] <... openat resumed> )      = 15
^C

real	0m8.692s
user	0m0.033s
sys	0m0.028s
```

**Comment by tbenz9** [comment]
Possibly related to #2215

**Comment by hackneyb** [comment]
I've seen the same behaviour but took it to be normal for this software.  Only concern for me was that it failed (once, access denied due to a lock) and logged an error in siad console, but I learned that that in itself didn't seem to cause an issue

I concur that constant IO could be an issue but it may be cached and not actually causing physical IO

**Comment by pablos91** [comment]
will this one be fixed eventually ? I'm checking it from the 1.3.0 version in 1.3.1 there's no improvement. I own a NAS in my room which is getting loud when constantly writing unneccessary things on disk. That's the reason I still didn't start my host and sold all my SC :<

**Comment by mckelvin** [comment]
I'm running `siad` on Linux, using a `ext4` disk as data dir. Everytime when I start the `siad`, I can see a process named `jbd2/sda1-8` will take nearly 100% of the IO (In `sudo iotop`) but it doesn't read or write. The jbd2 is a ext4 related process. I think it's relevant to this issue.

```
  TID  PRIO  USER     DISK READ  DISK WRITE  SWAPIN     IO>    COMMAND
  815 be/3 root        0.00 B/s    0.00 B/s  0.00 % 99.99 % [jbd2/sda1-8]
```

**Comment by mckelvin** [comment]
# A workaround for iotop IO 100%

@Fornax96 told me on the Discord channel that the `siad` is always using nearly 100% disk I/O. Before that I thought it's just introduced in the latest release.

In order to save my disk. I did a trick by move all directories except `consensus` (which is big) to memory (`/dev/shm`) and link back to the original sia data dir.

```
ubuntu@ubuntu:/siad-data$ ls -al
total 16
drwxr-xr-x 3 ubuntu root   4096 Jan  9 22:53 .
drwxr-xr-x 3 ubuntu root   4096 Jan  9 20:24 ..
drwx------ 2 ubuntu ubuntu 4096 Jan  9 20:25 consensus
lrwxrwxrwx 1 ubuntu ubuntu   26 Jan  9 22:53 gateway -> /dev/shm/siad-data/gateway
lrwxrwxrwx 1 ubuntu ubuntu   23 Jan  9 22:53 host -> /dev/shm/siad-data/host
lrwxrwxrwx 1 ubuntu ubuntu   25 Jan  9 22:53 renter -> /dev/shm/siad-data/renter
lrwxrwxrwx 1 ubuntu ubuntu   34 Jan  9 22:53 transactionpool -> /dev/shm/siad-data/transactionpool
lrwxrwxrwx 1 ubuntu ubuntu   25 Jan  9 22:53 wallet -> /dev/shm/siad-data/wallet
```

The bad thing of this workaround is I have to backup the data in `/dev/shm/siad-data` before the server is shutdown.

**Comment by lukechampine** [comment]
@mckelvin can you run this command to identify which file is being accessed most frequently?
```
inotifywait -e modify -e move -e create -e delete -m -r [siad directory]
```
Also try running with this patch: https://github.com/NebulousLabs/Sia/pull/2362

**Comment by mckelvin** [comment]
@lukechampine I start `inotifywait -e modify -e move -e create -e delete -m -r /siad-data | tee -a sia_io.log`, and here is the stats:

```
$ sort siad_io.log  | uniq -c
  85829 /siad-data/consensus/ MODIFY consensus.db
     14 /siad-data/gateway/ MODIFY nodes.json
     16 /siad-data/gateway/ MODIFY nodes.json_temp
   1027 /siad-data/host/contractmanager/ CREATE contractmanager.json_temp
   1027 /siad-data/host/contractmanager/ CREATE contractmanager.wal_temp
   1685 /siad-data/host/contractmanager/ MODIFY contractmanager.json_temp
   1721 /siad-data/host/contractmanager/ MODIFY contractmanager.wal_temp
   1027 /siad-data/host/contractmanager/ MOVED_FROM contractmanager.json_temp
   1027 /siad-data/host/contractmanager/ MOVED_FROM contractmanager.wal_temp
   1027 /siad-data/host/contractmanager/ MOVED_TO contractmanager.json
   1027 /siad-data/host/contractmanager/ MOVED_TO contractmanager.wal
   1132 /siad-data/host/ MODIFY host.db
   1043 /siad-data/host/ MODIFY host.json
   1062 /siad-data/host/ MODIFY host.json_temp
   1046 /siad-data/renter/ MODIFY contractor.json
   1041 /siad-data/renter/ MODIFY contractor.json_temp
     13 /siad-data/renter/ MODIFY hostdb.json
     15 /siad-data/renter/ MODIFY hostdb.json_temp
      1 /siad-data/renter/ MODIFY renter.log
  11055 /siad-data/transactionpool/ MODIFY transactionpool.db
     17 /siad-data/wallet/ MODIFY wallet.db
```

Since I'm sync'ing the blockchain (progress: 75%) right now. It's reasonable that `consensus.db` is the most frequent one.

NOTE: Sometimes IO of `jbd2/sda1-8` is very high, sometimes IO of`siad` is very high. Either of this will only happen when the `siad` is started. Even though the IO rate is very high, the read/write are very small.

**Comment by mckelvin** [comment]
After applying #2362, which doesn't make big difference for IO util%. I verified twice to make sure the patch is already applied and I was using the correct siad binary.

Here is the corresponding output of inotifywait
```
$ sort siad_io.log  | uniq -c
  58376 /siad-data/consensus/ MODIFY consensus.db
      4 /siad-data/consensus/ MODIFY consensus.log
     44 /siad-data/gateway/ MODIFY nodes.json
     43 /siad-data/gateway/ MODIFY nodes.json_temp
    672 /siad-data/host/ MODIFY host.db
    640 /siad-data/host/ MODIFY host.json
    640 /siad-data/host/ MODIFY host.json_temp
    646 /siad-data/renter/ MODIFY contractor.json
    628 /siad-data/renter/ MODIFY contractor.json_temp
      8 /siad-data/renter/ MODIFY hostdb.json
      8 /siad-data/renter/ MODIFY hostdb.json_temp
   6873 /siad-data/transactionpool/ MODIFY transactionpool.db
     10 /siad-data/wallet/ MODIFY wallet.db
```

**Comment by lukechampine** [comment]
the `contractmanager.wal` and `contractmanager.json` files are clearly being accessed less frequently. I'd check again after you finish syncing consensus.db.

**Comment by hackneyb** [comment]
I can reproduce the same issue as I posted before (running Windows, same behavior).  Just not convinced it's a bug.  My IO is high in two ways: temp files on json files (constant renaming) and zeroing out the storage file when renter uploads data.  My second issue is detailed here: https://github.com/NebulousLabs/Sia/issues/2594

**Comment by Fnordy** [comment]
I'm also seeing high IO running 1.3.1 in a docker container on a Synology NAS, which isn't doing anything else.  Its just taken me near a week to sync while "volume" utilization is up at 99%.  If i stop siad volume util drops to 0%.   

Volume util dropped to around 30% after the blockchain sync completed but was still higher than it ought to be for doing nothing! (I can hear the NAS drives working furiously - but its only running siad!).

The above commands to determine which files are used don't work on the NAS CLI - so can't easily provide that.

This High IO with no actual actiivty makes me too concerned to use SIA full time... any suggestions?

**Comment by lukechampine** [comment]
@Fnordy please try applying patch https://github.com/NebulousLabs/Sia/pull/2362. It should greatly reduce the IO.

**Comment by Fnordy** [comment]
Ta for reply - but Dang; figuring out a go & a re-compile will take me a while.  May need to wait until next release.  If I get time to have a longer play I'll let you know!

**Comment by chargr** [comment]
@lukechampine I have applied the patch over 1.3.1 and it is much improved. Now peaking at 90 write ops/sec and idling around 40 write ops/sec (i'd be 10-20 without sia running).  iowait is down (gone). Thanks.

@Fnordy this is my method (assuming you have already run go get)
```
git checkout v1.3.1
wget https://github.com/NebulousLabs/Sia/pull/2362.patch
patch -p1 < 2362.patch
mkdir out
go build -a -v -tags 'netgo' -ldflags='-s -w' -o out/siac ./cmd/siac
go build -a -v -tags 'netgo' -ldflags='-s -w' -o out/siad ./cmd/siad
sudo install out/siac /path/to/your/sia/bin/siac
sudo install out/siad /path/to/your/sia/bin/siad
```

**Comment by Fnordy** [comment]
@chargr Great stuff, thanks for that - copied exact process and I see the same thing - you may have saved a tree or two. :)  Odd write ops is still so high, but I'll take it :)

**Comment by tbenz9** [comment]
Fixed in #2362
